src/lib/misc/roughtime/roughtime.h

```diff
@@ -0,0 +1,119 @@
+
+namespace Roughtime {
+
+constexpr unsigned request_min_size = 1024;
+
+/**
+* An Roughtime request.
+*/
+BOTAN_PUBLIC_API(2,10)
+std::array<uint8_t, request_min_size> encode_request(const std::array<uint8_t, 64>& nonce);
+
+/**
+* An Roughtime response.
+*/
+class BOTAN_PUBLIC_API(2,10) Response final
+   {
+   public:
+      using sys_microseconds64 = std::chrono::time_point<std::chrono::system_clock, std::chrono::duration<uint64_t,std::micro>>;
+      using microseconds32 = std::chrono::duration<uint32_t,std::micro>;
+
+      static Response from_bits(const std::vector<uint8_t>& response, const std::array<uint8_t, 64>& nonce);
+
+      bool validate(const Ed25519_PublicKey& pk) const;
+
+      sys_microseconds64 utc_midpoint() const { return m_utc_midpoint; }
+
+      microseconds32 utc_radius() const { return m_utc_radius; }
+   private:
+      Response(std::array<uint8_t,72> dele, std::array<uint8_t,64> sig, sys_microseconds64 utc_midp, microseconds32 utc_radius)
+      : m_cert_dele (dele)
+      , m_cert_sig (sig)
+      , m_utc_midpoint {utc_midp}
+      , m_utc_radius {utc_radius}
+      {}
+      const std::array<uint8_t,72> m_cert_dele;
+      const std::array<uint8_t,64> m_cert_sig;
+      const sys_microseconds64 m_utc_midpoint;
+      const microseconds32 m_utc_radius;
+   };
+
+struct BOTAN_PUBLIC_API(2,10) Link final
+   {
+      std::vector<uint8_t> response;
+      Ed25519_PublicKey public_key;
+      std::array<uint8_t, 64> nonce_or_blind;
+   };
+
+class BOTAN_PUBLIC_API(2,10) Chain final
+   {
+   public:
+      Chain() = default; //empty
+      Chain(DataSource& ds);
+      const std::vector<Link>& links() const { return m_links; }
+      std::vector<Response> responses() const;
+      void next_blind_and_nonce(RandomNumberGenerator& rng,
+                                std::array<uint8_t, 64>& nonce,
+                                std::array<uint8_t, 64>& blind) const;
+      void append(const std::vector<uint8_t>& response,
+                  const Ed25519_PublicKey& public_key,
+                  const std::array<uint8_t, 64>& nonce_or_blind,
+                  size_t max_chain_size);
+      void persist(DataSink_Stream& ds) const;
+   private:
+      std::vector<Link> m_links;
+   };
+
+/**
+*/
+BOTAN_PUBLIC_API(2,10)
+std::array<uint8_t, 64> nonce_from_blind(const std::vector<uint8_t>& previous_response,
+                                         const std::array<uint8_t, 64>& blind);
+
+/**
+* Makes an online Roughtime request via UDP and returns the Roughtime response.
+* @param url Roughtime server UDP endpoint (host:port)
+* @param timeout a timeout on the UDP request
+* @return Roughtime response
+*/
+BOTAN_PUBLIC_API(2,10)
+std::vector<uint8_t> online_request(const std::string& uri,
+                        const std::array<uint8_t, 64>& nonce,
+                        std::chrono::milliseconds timeout = std::chrono::seconds(3));
+
+
+struct BOTAN_PUBLIC_API(2,10) Server final
+  {
+    std::string name;
```

Comment:
Avoid public members: instead use `class`, a constructor, and getters returning `const &` for the various members. This adds boilerplate but prevents modifications of a `Server` once deserialized.